authentication/oidc.go

```diff
@@ -45,71 +47,112 @@ type OIDCConfig struct {
 // Middleware is a convenience type for functions that wrap http.Handlers.
 type Middleware func(http.Handler) http.Handler
 
-// NewOIDC creates a single http.Handler and a set of Middlewares for all
+// OIDCHandlers holds handlers and a set of middlewares for all
 // tenants that is able to authenticate requests and provide the
 // authorization code grant flow for users.
-func NewOIDC(
-	logger log.Logger,
-	reg prometheus.Registerer,
-	prefix string,
-	configs []TenantOIDCConfig,
-) (http.Handler, map[string]Middleware, []error) {
-	handlers := map[string]http.Handler{}
-	middlewares := map[string]Middleware{}
-	warnings := make([]error, 0, len(configs))
+type OIDCHandlers struct {
+	mtx         sync.RWMutex
+	handlers    map[string]http.Handler
+	middlewares map[string]Middleware
 
+	logger          log.Logger
+	retryCountIncFn func()
+}
+
+// NewOIDCHandlers instantiates OIDC handlers.
+func NewOIDCHandlers(l log.Logger, reg prometheus.Registerer) *OIDCHandlers {
 	tenantErrors := prometheus.NewGauge(prometheus.GaugeOpts{
 		Namespace: "observatorium",
 		Subsystem: "api",
 		Name:      "tenant_registration_failures",
-		Help:      "The number of tenants for which OIDC provider instantiation failed.",
+		Help:      "The number of attempts for which tenant OIDC provider instantiation failed.",
 	})
 
 	reg.MustRegister(tenantErrors)
 
-	for _, c := range configs {
-		p, err := NewProvider(context.TODO(), logger, getCookieForTenant(c.Tenant), "/"+c.Tenant, c.OIDCConfig)
-		if err != nil {
-			warnings = append(warnings, fmt.Errorf("failed to instantiate OIDC provider for tenant %q: %w", c.Tenant, err))
+	return &OIDCHandlers{
+		handlers:        make(map[string]http.Handler),
+		middlewares:     make(map[string]Middleware),
+		logger:          l,
+		retryCountIncFn: tenantErrors.Inc,
+	}
+}
 
-			tenantErrors.Inc()
+// AddOIDCForTenant adds a new handler and middleware for a new tenant, based on
+// the tenant OIDC configuration. If a connection with the OIDC provider
+// cannot be established, the method will retry to connect with an increasing
+// backoff.
+func (oh *OIDCHandlers) AddOIDCForTenant(prefix string, config TenantOIDCConfig) {
+	ctx := context.Background()
+	b := backoff.New(ctx, backoff.Config{
+		Min:        500 * time.Millisecond,
+		Max:        5 * time.Second,
+		MaxRetries: 0, // retry indefinitely
+	})
 
-			continue
-		}
+	go func() {
+		for b.Reset(); b.Ongoing(); {
+			p, err := NewProvider(ctx, oh.logger, getCookieForTenant(config.Tenant), "/"+config.Tenant, config.OIDCConfig)
+			if err != nil {
+				level.Warn(oh.logger).Log("msg", "failed to instantiate OIDC provider for tenant", "tenant", config.Tenant, "error", err)
+				oh.retryCountIncFn()
+				b.Wait()
+				continue
+			}
 
-		r := chi.NewRouter()
+			r := chi.NewRouter()
```

Comment:
nit: There are things here that probably _should_ be instantiated outside of the for loop?

Reply:
Which ones do you mean? Since the loop will progress to this section only once, I think it should be fine even if it is instantiated here. From the point of code clarity, I'm impartial whether it is more 'reasonable' to do it within or outside.

Reply:
I think there some reasons why it can be more readable if we make loop more concise and smaller, I was surprised too here, same as @ianbillett . Reasons:

* Two engineers were already surprised by this part 🙈 
* If you do it in loop I expect this code to use loop functionality, as is - that we can go back to retries, but looks this large part, does cannot. Focused scoped complexity is the key
* Less indent for complex code = better.

Reply:
Ok now I'm convinced :upside_down_face:

Reply:
I did move the new router instantiation, as far as I could see that was actually the only thing being instantiated in the loop that could be done outside :slightly_smiling_face: .